src/braket/circuits/box_drawing_circuit_diagram.py

```diff
@@ -0,0 +1,524 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+from __future__ import annotations
+
+from functools import reduce
+from typing import Literal, Union
+
+import braket.circuits.circuit as cir
+from braket.circuits.circuit_diagram import CircuitDiagram
+from braket.circuits.compiler_directive import CompilerDirective
+from braket.circuits.gate import Gate
+from braket.circuits.instruction import Instruction
+from braket.circuits.moments import MomentType
+from braket.circuits.noise import Noise
+from braket.circuits.result_type import ResultType
+from braket.registers.qubit import Qubit
+from braket.registers.qubit_set import QubitSet
+
+
+class BoxDrawingCircuitDiagram(CircuitDiagram):
```

Comment:
I would imagine the high-level logic for constructing these diagrams is quite similar to `AsciiCircuitDiagram`. Can we share more logic between the two classes? Maybe even make `BoxxDrawingCircuitDiagram` a subclass of `AsciiCircuitDiagram` and factor out common logic into helper methods? That would help maintainability significantly